### ISSUE TITLE
Remote service discovery fails if the local Consul server cluster loses quorum

### ISSUE THREAD

**Issue by dougbarth-stripe** [hashicorp/consul#3630]
If you have a question, please direct it to the
[consul mailing list](https://www.consul.io/community.html) if it hasn't been
addressed in either the [FAQ](https://www.consul.io/docs/faq.html) or in one
of the [Consul Guides](https://www.consul.io/docs/guides/index.html).

When filing a bug, please include the following:

### `consul version` for both Client and Server
Client: `0.9.3` (not relevant in this case)
Server: `0.9.3`

### `consul info` for both Client and Server
Client:
```
[Client `consul info` here]
```

Server:
```
agent:
	check_monitors = 2
	check_ttls = 0
	checks = 2
	services = 5
build:
	prerelease =
	revision = 112c060
	version = 0.9.3
consul:
	bootstrap = false
	known_datacenters = 2
	leader = false
	leader_addr = 10.xx.xx.215:8300
	server = true
raft:
	applied_index = 7798727
	commit_index = 7798727
	fsm_pending = 0
	last_contact = 19.825862ms
	last_log_index = 7798727
	last_log_term = 2786
	last_snapshot_index = 7797741
	last_snapshot_term = 2786
	latest_configuration = [{Suffrage:Voter ID:0b182968-SNIP Address:10.xx.xx.78:8300} {Suffrage:Voter ID:edabfd71-SNIP Address:10.xx.xx.215:8300} {Suffrage:Voter ID:f44b2b3e-SNIP Address:10.xx.xx.237:8300}]
	latest_configuration_index = 5064475
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 2786
runtime:
	arch = amd64
	cpu_count = 4
	goroutines = 275
	max_procs = 4
	os = linux
	version = go1.9
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 69
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 6728
	members = 56
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 15122
	members = 6
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details
Ubuntu 14.04/16.04
AWS multiple regions

### Description of the Issue (and unexpected/desired result)
We run a system where clients discover Consul services via Consul's DNS interface. Instead of having clients query their local Consul agent, we run a set of DNS servers running unbound. The unbound recursors are configured to forward requests to our Consul servers' agent on the DNS interface. The Consul servers are configure to allow stale results to be returned, so we prefer inconsistent results in the face of network partitions. We run multiple datacenters and our Consul servers are configured to allow cross datacenter requests

We ran a test were we introduced artificial network partitions into one datacenter's consul server cluster while querying for services in the local datacenter and the remote datacenter.

We found that Consul servers are unable to make RPC requests to a remote datacenter when enough nodes are network isolated to lose quorum in the local datacenter. When attempting to resolve records in the remote datacenter using Consul's DNS interface in the local datacenter, Consul servers that are not network isolated will return a SERVFAIL result. Queries for local records succeed.

The `consul operator raft list-peers` command works in the local datacenter when run on the non-isolated Consul servers (it hangs on isolated servers), but for the remote datacenter it returns `No path to datacenter` for both network isolated and non-isolated servers.

We allow stale reads for consul services via the DNS interface. We would expect that queries for remote datacenter services would succeed if requested via non-isolated Consul servers.

### Reproduction steps
Create two Consul clusters in different datacenters with each datacenter having 3 server nodes and all server nodes joined to the WAN gossip pool.

In one datacenter, issue requests for a local service and a remote service continuously.

Network isolate a single Consul server using the following commands (under Linux):

```
sudo iptables -I INPUT 1 -p tcp --dport 8302 -j DROP
sudo iptables -I INPUT 1 -p udp --dport 8302 -j DROP

sudo iptables -I OUTPUT 1 -p tcp --sport 8302 -j DROP
sudo iptables -I OUTPUT 1 -p udp --sport 8302 -j DROP

sudo iptables -I INPUT 1 -p tcp --dport 8301 -j DROP
sudo iptables -I INPUT 1 -p udp --dport 8301 -j DROP

sudo iptables -I OUTPUT 1 -p tcp --sport 8301 -j DROP
sudo iptables -I OUTPUT 1 -p udp --sport 8301 -j DROP

sudo iptables -I INPUT 1 -p tcp --dport 8300 -j DROP
sudo iptables -I INPUT 1 -p udp --dport 8300 -j DROP

sudo iptables -I OUTPUT 1 -p tcp --sport 8300 -j DROP
sudo iptables -I OUTPUT 1 -p udp --sport 8300 -j DROP

sudo iptables -I INPUT 1 -p tcp --dport 8600 -j DROP
sudo iptables -I INPUT 1 -p udp --dport 8600 -j DROP

sudo iptables -I OUTPUT 1 -p tcp --sport 8600 -j DROP
sudo iptables -I OUTPUT 1 -p udp --sport 8600 -j DROP
```

Queries against the DNS interface on the isolated server should fail, but for non-isolated servers should succeed.

Isolating a second server (thus losing quorum) should result in local services resolving on non-isolated servers and failing on isolated servers. Remote service discovery fails on both isolated and non-isolated services.

Note: This issue is written based off my ad hoc tests. I'm happy to work up more rigorous results if needed.

**Comment by dougbarth-stripe** [comment]
I tried recreating this issue in a Docker environment, but I was unsuccessful. I was similarly unable to reproduce this issue in our test environment, so it appears that my original theory is invalid. I'm going to close this, but I'd love any ideas to check for what might have caused cross DC RPCs to fail.

**Comment by slackpad** [comment]
Hi @dougbarth-stripe, https://github.com/hashicorp/consul/issues/3463 could cause some weird issues with RPCs where you could get stuck on a dead server for some time. That was fixed in 1.0.1.
